src/sentry/south_migrations/0513_fix_gh_bb_repo_data.py

```diff
@@ -0,0 +1,1700 @@
+
+    # Flag to indicate if this migration is too risky
+    # to run online and needs to be coordinated for offline
+    is_dangerous = False
+
+    def forwards(self, orm):
+        db.commit_transaction()
+        try:
+            self._forwards(orm)
+        except Exception:
+            db.start_transaction()
+            raise
+        db.start_transaction()
+
+    def _forwards(self, orm):
+        """
+        Before https://github.com/getsentry/sentry/pull/12290, for GitHub and
+        BitBucket repos we were only updating `repo.config['name']` when repo data
+        in webhook events was different than what we had, while *not* updating
+        `repo.name` or `repo.url`.
+
+        Since that was merged, webhooks have fixed some data but not all. This
+        updates all GitHub or BitBucket repos so that all data matches `config['name]`.
+
+        See comment below for testing.
+        """
+
+        # functions, etc for annotating, filtering, and fixing Repository records
+
+        # we store the config as a string, so first cast it to JSON, then grab
+        # the 'name' value
+        config_as_json = Func(F('config'), function='json')
+        config_name = Func(config_as_json, Value('name'), function='json_extract_path_text')
+
+        # strip the scheme and hostname (url_prefix) from the url
+        url_prefix = Func('url', Value('https?://[^/]*/'), function='substring')
+        url_path = Func('url', url_prefix, Value(''), function='replace')
+
+        # create the correct url using config_name
+        fixed_url = Func(url_prefix, config_name, function='concat')
+
+        # we want to test for A = B = C, so let's use the transitive property
+        name_equals_path = Q(name__notdistinctfrom='url_path')
+        path_equals_config_name = Q(url_path__notdistinctfrom='config_name')
+
+        # now actually do the work
+
+        # grab only GitHub and BitBucket repos, because we don't want to be
+        # changing things which aren't the result of the bug
+        repos = orm.Repository.objects.filter(provider__in=[
+            'github',
+            'integrations:github',
+            'bitbucket',
+            'integrations:bitbucket'
+        ])
+
+        # add `config_name` and `url_path` from above as temporary fields in the
+        # table, so as to be able to use them later
+        repos = repos.annotate(config_name=config_name, url_path=url_path)
+
+        # git rid of any repos whose data *isn't* corrupted, in other words, any
+        # repos where `name`, `config_name`, and `url_path` all match
+        repos = repos.exclude(name_equals_path & path_equals_config_name)
+
+        # also get rid of any which don't have a value for config['name'] so we
+        # don't overwrite existing data with NULLs
+        repos = repos.exclude(config_name__isnull=True)
+
+        # finally, now that we only have broken repos, fix the data
+        repos.update(name=config_name, url=fixed_url)
```

Comment:
How many rows is this updating? If it is more than a few hundred this should probably be an `is_dangerous` flag so that operations can run it post-deploy.

Reply:
1388 as of 2pm Thursday. (The number drops periodically as repos get fixed by the webhooks.) Sounds from your definition like I should flip it to dangerous, yes?

Reply:
I would run this out of band from a deploy personally.